Support V2 GRPC for KServe Model Server

Suresh-Nakkeran opened the pull request:
**What this PR does / why we need it**:
Add gRPC support for KServe custom model server

**Which issue(s) this PR fixes**:
Fixes #2107 

**Type of changes**
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:
- [x] gRPC custom predictor
- [x] gRPC custom transformer with gRPC predictor
- [x] HTTP custom transformer with gRPC predictor

**Special notes for your reviewer**:


**Release note**:
```
Enable gRPC model server with `--enable_grpc`,  by default it is true.
Configure max gRPC processing threads with `--max_threads`, by default it is 4.
```

yuzisun commented on python/custom_model/model_grpc.py:
```diff
@@ -0,0 +1,98 @@
+# Copyright 2021 The KServe Authors.
```
```suggestion
# Copyright 2022 The KServe Authors.
```

yuzisun commented on python/kserve/kserve/grpc/grpc_predict_v2.proto:
```diff
@@ -0,0 +1,362 @@
+// Copyright 2020 kubeflow.org.
```
```suggestion
# Copyright 2022 The KServe Authors.
```

yuzisun commented on docs/samples/kafka/image_transformer/__main__.py:
```diff
@@ -27,4 +28,6 @@
     transformer = ImageTransformer(args.model_name, predictor_host=args.predictor_host)
     server = kserve.ModelServer()
-    server.start(models=[transformer])
+    asyncio.run(
```
This seems like a breaking change for user, can we do `asyncio` inside `server.start`?

yuzisun replied:
So you can't execute `asyncio.run` inside the `server.start`?

yuzisun commented on test/e2e/custom/test_custom_model_grpc.py:
```diff
@@ -0,0 +1,78 @@
+        containers=[
+            V1Container(
+                name="kserve-container",
+                image="suryasmart619/custom-model-grpc:v1", #TODO: Use Kserve repository
```
I think we can build the `CustomModel` as e2e test image like how did with transformer
https://github.com/kserve/kserve/blob/master/test/scripts/gh-actions/build-server-runtimes.sh#L67

yuzisun commented on python/kserve/kserve/model.py:
```diff
@@ -169,9 +170,7 @@ def postprocess(self, response: Union[Dict, ModelInferResponse]) -> Dict:
         :param response: Dict|ModelInferResponse passed from predict handler
         :return: Dict
         """
-        if isinstance(response, ModelInferResponse):
```
need to check whether the original request is REST or gRPC by examining the headers

yuzisun commented on python/kserve/kserve/handlers/dataplane.py:
```diff
@@ -0,0 +1,62 @@
+import inspect
+
+from kserve.model_repository import ModelRepository
+from kserve.utils.utils import generate_uuid
```
not used

yuzisun commented on python/kserve/kserve/model_server.py:
```diff
@@ -136,7 +139,7 @@ def start(self, models: Union[List[Model], Dict[str, Deployment]], nest_asyncio:
             import nest_asyncio
             nest_asyncio.apply()
 
-        tornado.ioloop.IOLoop.current().start()
```
first start grpc server and then tornado

Suresh-Nakkeran committed:
Implemented gRPC in kserve

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

yuzisun commented on python/kserve/kserve/grpc/server.py:
```diff
@@ -0,0 +1,40 @@
+
+    def _create_server(self):
+        self._inference_servicer = InferenceServicer(self._data_plane)
+        self._server = aio.server(futures.ThreadPoolExecutor(max_workers=5))
```
why setting to 5? can we make it configurable?

Suresh-Nakkeran replied:
I will update with param max_asyncio_workers.

yuzisun commented on python/kserve/kserve/grpc/server.py:
```diff
@@ -0,0 +1,40 @@
+        )
+        await self._server.wait_for_termination()
+
+    async def stop(self, sig: int = None):
```
`sig` is not used here

yuzisun commented on test/e2e/custom/test_custom_model_grpc.py:
```diff
@@ -0,0 +1,90 @@
+        containers=[
+            V1Container(
+                name="kserve-container",
+                image="suryasmart619/custom-model-grpc:v1",  # TODO: Use Kserve repository
```
We can build the custom model image in kserve CI

yuzisun commented on python/kserve/kserve/utils/utils.py:
```diff
@@ -116,3 +118,20 @@ def create_response_cloudevent(model_name: str, body: Union[Dict, CloudEvent], r
+
+def to_headers(context: ServicerContext) -> Dict[str, str]:
+    metadata = context.invocation_metadata()
+    if hasattr(context, "trailing_metadata"):
```
what's `trailing_metadata` ?

yuzisun commented on .github/workflows/python-test.yml:
```diff
@@ -23,7 +23,7 @@ jobs:
         with:
           max-line-length: "120"
-          exclude: "docs/samples/v1beta1/onnx/assets/*.py,python/kserve/test/test_v1*.py,python/kserve/kserve/__init__.py,python/kserve/test/test_knative*.py"
+          exclude: "docs/samples/v1beta1/onnx/assets/*.py,python/kserve/test/test_v1*.py,python/kserve/kserve/__init__.py,python/kserve/test/test_knative*.py,python/kserve/kserve/grpc/grpc_predict_v2*.py"
```
Why excluding the grpc test?

Suresh-Nakkeran replied:
@yuzisun  Added to skip lint issue with proto files

yuzisun commented on python/custom_model/model_grpc.py:
```diff
@@ -0,0 +1,98 @@
+                    "data": result[0],
+                    "datatype": "FP32",
+                    "name": req.name,
+                    "shape": req.shape
```
Here should use the response's data shape

yuzisun commented on python/custom_model/model_grpc.py:
```diff
@@ -0,0 +1,98 @@
+        elif isinstance(request, ModelInferRequest):
+            req = request.inputs[0]
+            fields = req.contents.ListFields()
+            _, field_value = fields[0]
```
we can directly access the `bytes_contents` field here?

yuzisun committed:
Fix gRPC custom model example and test

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun commented on python/kserve/kserve/model_server.py:
```diff
@@ -155,7 +158,7 @@ def start(self, models: Union[List[Model], Dict[str, Deployment]], nest_asyncio:
             nest_asyncio.apply()
 
-        tornado.ioloop.IOLoop.current().start()
+        await self._grpc_server.start(self.max_asyncio_workers)
```
use `self.workers`

yuzisun commented on python/kserve/kserve/model_server.py:
```diff
@@ -83,6 +84,8 @@ def __init__(self, http_port: int = args.http_port,
 
     def create_application(self):
+        dataplane = handlers.DataPlane(model_registry=self.registered_models)
+        self._grpc_server = GRPCServer(port=self.grpc_port, data_plane=dataplane)
```
`create_application` is tornado specific, it is better to separate out grpc server related code.

yuzisun committed:
Add python v2 grpc type interface file

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Fix transformer gRPC example

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

Suresh-Nakkeran committed:
Fix custom grpc docker file

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

yuzisun committed:
Merge branch 'master' into feature/kserve-grpc-custom-model

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

Suresh-Nakkeran committed:
Merge branch 'master' into feature/kserve-grpc-custom-model

Suresh-Nakkeran committed:
Fixed issue in master merge
Added kserve version in custom_model_grpc Dockerfile
Added ServerMetadata function in grpc servicer
Truncated points in grpc test case response

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

Suresh-Nakkeran committed:
Removed dependency of asyncio in server start

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

alexagriffith commented on python/custom_model/runtime.txt:
```diff
@@ -0,0 +1 @@
+python-3.8.14
```
this is python 3.8 but the image builds python 3.9, should we keep consistent?

alexagriffith commented on python/kserve/kserve/grpc/grpc_predict_v2.proto:
```diff
@@ -0,0 +1,362 @@
+// Copyright 2022 The KServe Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
```
just curious, why is this different from https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/grpc_predict_v2.proto ?

yuzisun replied:
The one in docs directory does not include the model repository extension for load/unload apis

ghost commented:
@yuzisun one thing i just realized is that we didn't put `fastapi` dependency in the `requirements.txt` file. right now the CI could pass because `ray[serve]` uses `fastapi`, I think we should be more explicit and add it to the `requirements.txt`

```
fastapi>=0.85.0
uvicorn>=0.16.0
```

yuzisun committed:
Fix asyncio.run for examples

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Add processing/asyncio threading arguments

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Gracefully terminate grpc server

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

Suresh-Nakkeran committed:
Merge branch 'master' into feature/kserve-grpc-custom-model

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

Suresh-Nakkeran committed:
Added support to read headers

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

Suresh-Nakkeran committed:
Added grpc response conversation

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

Suresh-Nakkeran committed:
Added logging interceptor in gRPC

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

Suresh-Nakkeran committed:
Added test cases for grpc model with transformer

Signed-off-by: Suresh-Nakkeran <suresh.n@ideas2it.com>

yuzisun commented:
Awesome work @Suresh-Nakkeran !!

/lgtm
/approve
